text/1013-datetime-arithmetic.rst

```diff
@@ -0,0 +1,226 @@
+          SELECT ("l" - "r")::int8
+      $$;
+  };
+
+The result is expected to be the number of days.
+
+
+Duration functions
+------------------
+
+We also will introduce normalization functions for ``relative_duration``::
+
+CREATE FUNCTION
+cal::normalize_relative_hours(dur: cal::relative_duration)
+  -> cal::relative_duration
+{
+    USING SQL FUNCTION 'justify_hours';
+};
+
+CREATE FUNCTION
+cal::normalize_relative_days(dur: cal::relative_duration)
+  -> cal::relative_duration
+{
+    USING SQL FUNCTION 'justify_days';
+};
+
+CREATE FUNCTION
+cal::normalize_relative_duration(dur: cal::relative_duration)
+  -> cal::relative_duration
+{
+    USING SQL FUNCTION 'justify_interval';
+};
+
+The ``normalize_relative_hours`` converts 24 hrs chunks into days.
+The ``normalize_relative_days`` converts days into months assuming that 1 month = 30 days.
+The ``normalize_relative_duration`` is the equivalent of calling
+``normalize_relative_days(normalize_relative_hours(x))``.
+
+Notice that for ``relative_duration`` the assumtion that 1 day is always 24
+hours holds true because it represents the time adjustment necessary on some
+clock rather than real time (like ``duration``). Therefore the
+``normalize_relative_hours`` transformation is technically lossless and could
+safely be reversed. However, not every month is 30 days and so the
+``normalize_relative_days`` and ``normalize_relative_duration`` are
+potentially lossy transformations aiming to produce an approximately
+equivalent ``relative_duration`` using months, years, etc. so care must be
+taken when using these conversions.
+
+We also need a function for extracting various ``duration`` and
+``relative_duration`` components::
+
+CREATE FUNCTION
+std::duration_get(dur: cal::relative_duration, el: std::str) -> std::float64
+{
+  ...
+};
+
+CREATE FUNCTION
+std::duration_get(dur: std::duration, el: std::str) -> std::float64
+{
+  ...
+};
+
+The components avaialable for extraction from ``relative_duration`` are:
+*millenium*, *century*, *decade*, *year*, *quarter*, *month*, *day*, *hour*,
+*minutes*, *seconds*, *milliseconds*, *microseconds*, and *totalseconds*. The
+*totalseconds* converts a ``relative_duration`` to seconds represented as a
+``float64`` value.
+
+
+Backwards Compatibility
+=======================
+
+This change is backwards-compatible. However, it is desirable to later
+deprecate arithmetic operators that allow interactions between the "local"
+date/time types and absolute ``duration`` as that is not well-defined.
+
+
+Security Implications
+=====================
+
+There are no security implications.
+
+
+Rejected Alternative Ideas
+==========================
+
+We specifically rejected the option of ``-`` operating on two ``local_date``
+values producing a ``relative_duration`` by default. In practice, this
+operation would produce ``relative_duration`` expressed using exclusively days
+anyways. The main difference then would be that in order to use the result of
```

Comment:
Why does it produce days only and not months or years? Is it artifact of the postgres implementation?

Reply:
Also even if it does like that, I'm still thinking that  `local_date - local_date = days` isn't obvious. One could assume it produces seconds or duration type as well depending on their background. So returning `relative_duration` with days only makes a lot of sense making code like `duration_get(date1 - date2,
'days')` self-documenting.

Reply:
> Why does it produce days only and not months or years? Is it artifact of the postgres implementation?

Because for date arithmetic `days` are exact, whereas months/years/etc. are not. It is also the case that Postgres implements it this way.

Perhaps we want to expose `age` function from Postgres that performs the subtraction exactly and returns the result using days, months, years, etc.

> So returning `relative_duration` with days only makes a lot of sense making code like `duration_get(date1 - date2, 'days')` self-documenting.

There's a performance penalty if you actually want days as an integer: `<int64>duration_get(date1 - date2, 'days')`. It's also pretty bulky for a common use-case.

Another reason relates to future plans for `range_unpack`. We want to be able to generate a bunch of dates from a range. By default we want to step by 1 day, but in principle we also want to allow other steps, like 2 days, 7 days, etc. The key here is that it's always a whole number of days, not fractional. So we've had this idea that it's really the `-` operator that defines the type of the step (as it does for all other types - because the step is in whatever units the "difference" is measured). If we use `relative_duration` for step instead we need to either discard the fractional part (like it happens for performing `+`) or forbid it outright making it an error. Both seem rather awkward. So these thoughts led us to the idea that actually `local_date` is much more like a special kind of integer than it is like `datetime`.

Reply:
Yeah, the analogy here is integers vs floating point types.  `relative_duration` is fractional, so it'd be like returning a `float64` from `int64 - int64`, which is loss of type fidelity.  As a compromise we can consider adding a `relative_duration` subdomain that allows only `days` and is implicitly castable to `relative_duration` and can also be cast from/to ints.  This way you can still get nice typing and output but without the ambiguity of `relative_duration` or weirdness of pure `int64`.

Reply:
> There's a performance penalty if you actually want days as an integer: `<int64>duration_get(date1 - date2, 'days')`. It's also pretty bulky for a common use-case.

I think we can instruct compiler to optimize this in most cases. While it's bulky, I'm pretty sure it's not so common to optimize brevity over clarity.

-----

On the other hand, I somewhat agree with the "minus defines type of the step" argument.

Will I still be able to step by month or year with this definition? Is it applied only for default step, or does this mean that integer days is the only step type I can use?

Reply:
> Will I still be able to step by month or year with this definition? Is it applied only for default step, or does this mean that integer days is the only step type I can use?

Good point.

I think we should go with the suggestion by @elprans: introduce a new subtype of `relative_duration` that does not have the "fractional day" components. We can call it `date_duration`.